### ISSUE TITLE
Vehicle Skill should help you anticipate your vehicle's future position

### ISSUE THREAD

**Issue by bouchacha** [CleverRaven/Cataclysm-DDA#59337]
### Is your feature request related to a problem? Please describe.

Driving a vehicle in a 3D video game is usually very fun and predictable. Driving in CDDA is very often a hassle, especially around obstacles and tight spaces.

1. Rotating vehicles take on largely unpredictable shapes, and it's never obvious whether the new shape will collide with anything until it's too late.
2. The turn-based mechanism also prevents the player from diagramming an implicit arc. If someone sees a real-time animation of a rock being thrown, it's generally fairly intuitive to predict the anticipated arc based on observing just a snippet of the animation. This would not be the case if the thrown rock moved along its arc trajectory in a turn-based manner; a lot of information (speed, tempo, power, etc) would be lost when displayed in that manner.
3. The vehicle skill is more or less useless. The common advice is to turn off progression completely because it saps up everyone's focus without providing a discernible benefit.
4. Driving through a city in CDDA is a giant chore. Ideally you want to go fast, but the unpredictable nature of the vehicle mechanics make such an endeavor fraught and frustrating, because you'll often collide with things you don't expect to collide with. Going slow helps, a little, but still results in unpredictable results.

### Solution you would like.

The vehicle skill should give you a ghost image of where your vehicle is expected to be in the near future, and what trajectory it's expected to take. This will take into account the current speed and turning trajectory, and ideally it will highlight any potential conflicts. The ghost image prediction should gain significant accuracy as your vehicle skill levels up (and maybe you can finally have proficiencies for different vehicle classes e.g. truck vs bicycle).

If the [current state of autodrive](https://github.com/CleverRaven/Cataclysm-DDA/issues/53832) is any indication, the big problem with this solution is that it will be mathematically expensive. The ghost vehicle would need to be immediately updated after every change of action. Apparently this imposes a significant amount of overhead on autodriving, but maybe that's because autodriving has to consider a wide range of speed & actions? Perhaps the overhead will be manageable if the only prediction that is requested is based on static inputs.

### Describe alternatives you have considered.

Do nothing. Let driving remain a chore (outside of humongous deathmobiles that can ram through everything), and let the vehicle skill remain useless.

**Comment by PatrikLundell** [comment]
Vehicle skill isn't completely useless. A high skill reduces the risk of skids and improves the ability to recover from skids. Apart from that, I think there is a skill level at which the number of steering actions per turn is increased.
While it's annoying to have focus drain from the driving skill, it reaches the maximum level reasonably quickly, at which point it ceases to drain focus.

I don't see that a pure prediction would be particularly expensive. The cost from auto drive calculations come from the fact that it needs to try out many different possible routes in many different steps. Calculating a ghost zone through which the vehicle will pass through the next tick based on the current settings isn't much calculation (about the same as the actual movement will cost, assuming there are no collisions). Calculating the potential paths of the next turn is up to seven times that cost (three potential steering actions in each direction, plus straight ahead). The challenging parts would be to actually display the results in a non confusing manner. A simple single turn current settings prediction could presumably be displayed by a transparent overlay corridor tinted in some color that's visible against the background, with some indication of collisions (and no prediction of what will happen as a result of a collision, i.e. the corridor would just go straight through the obstruction with collision indications at the collision points).
An alternative to a visual display of the corridor could be some kind of turn collision indicator with 7 tiles indicating each of the up to 7 possible choices indicating Unavailable (because you don't have enough steering choices left to reach it), Collision (including the ones resulting from the vehicle smashing into obstacles on the sides due to morphing), and No Collision, e.g. like this "-[[<_>>]-", with color assisting ("-" = Unavailable, "[" = collision after left turn(s), "<" =clear left turn(s), "_" = clear straight ahead, "!" = collision straight ahead, ">" = clear right turn(s), "]" = collision after right turn(s)).
I think the alternative would be the one that's easiest to use.

**Comment by kevingranade** [comment]
This issue has been mentioned on **Cataclysm: Dark Days Ahead**. There might be relevant details there:

https://discourse.cataclysmdda.org/t/making-vehicle-skill-worthwhile-predictive-pathing/28039/1

**Comment by bouchacha** [comment]
> A high skill reduces the risk of skids and improves the ability to recover from skids.

Skids only happen with high-speed turns, right? Which means that recovering from a skid is only useful if you were turning near an obstacle. But the whole issue is that the current UI/control system discourages trying anything that risky except in wide-open areas.

> it reaches the maximum level reasonably quickly, at which point it ceases to drain focus.

This is technically true, but it just means that the downside eventually stops being a downside. Turning off the vehicle skill at the beginning feels like a power-gaming exploit, because it prevents having your focus (a very precious resource) get wasted on a skill with no real discernible benefits.

> I don't see that a pure prediction would be particularly expensive.

This is encouraging to hear! I really appreciate the detail you included in your response.

**Comment by PatrikLundell** [comment]
No. Skids can happen as a result of collisions with zombies and attacks by critters as well. This includes worms that sneak attack you on a seemingly empty plain. I think that at a low skill skids can happen when (trying to) moving straight too.

**Comment by Malkeus** [comment]
Id like the next turn ghost to be standard. The skill based predictive highlighting seems unnecessarily complex and sounds like there'd be a lot of overhead. 
The current 'losing control' mechanic that happens with low skill could be extended to speed control as well instead. It would definitely make you want to practice first if there was a risk of absolutely tromping on the accelerator when you meant to stomp on the brakes. 

I dunno about you lot, but the thing that tripped me up the most was achieving a steady acceleration/deceleration and maintaining a desired speed. It was all sudden acceleration followed by overcompensating on the brakes. This state didn't last long, but I'm sure I would have wrecked if I'd tried to go on a drive like that.

Driving really isn't that hard, in a modern car. Something you welded together out of scraps and liberated parts from disparate vehicles should be more of a chore, if only because it's hard to get the angles right with your eyeball and your steering alignment is off. Even a few degrees difference in alignment between the wheels can make your car pull or lessen your control. Loose tie rods cause a wobbly and meandering driving experience. 

Id like to see a system where driving skill of players was coupled with a driving difficulty score on vehicles. Something like a moped is absolutely going to be less difficult to operate than a semi truck. The current system treats both vehicles the same. Edit: maybe weight affects things? Not sure what the current system is looking at aside from skill level. Either way, it's incredibly quick and easy to level up to perfect driving and the skids at low levels hardly affect things unless your driving recklessly fast already.

Tldr: make the ghost standard, nix the predictive highlighting. Make skill matter by adding loss of speed control to skidding and introduce driving difficulty levels to vehicles.
